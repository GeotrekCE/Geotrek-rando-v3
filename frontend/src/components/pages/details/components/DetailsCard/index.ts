export { CardSingleImage, DetailsCard } from './DetailsCard';
