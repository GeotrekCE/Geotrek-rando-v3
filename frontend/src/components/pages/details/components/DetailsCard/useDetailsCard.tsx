import debounce from 'debounce';
import useIsomorphicLayoutEffect from 'hooks/useIsomorphicLayoutEffect';
import { useCallback, useEffect, useRef, useState } from 'react';

const DETAILS_CARD_DEFAULT_HEIGHT = 220;

export const useDetailsCard = (hasMedia = false) => {
  const detailsCardRef = useRef<HTMLDivElement>(null);
  const [truncateState, setTruncateState] = useState<'NONE' | 'TRUNCATE' | 'FULL'>('TRUNCATE');

  const toggleTruncateState = () =>
    setTruncateState(currentTruncateState =>
      currentTruncateState === 'TRUNCATE' ? 'FULL' : 'TRUNCATE',
    );

  useEffect(() => {
    const descriptionNode = detailsCardRef.current?.querySelector<HTMLElement>('.line-clamp-2');
    if (descriptionNode?.offsetHeight === descriptionNode?.scrollHeight && !hasMedia) {
      setTruncateState('NONE');
    }
  }, [hasMedia]);

  const handleResize = useCallback(
    debounce(
      () => {
        setTruncateState(prevState => {
          if (detailsCardRef.current === null || hasMedia) {
            return prevState;
          }
          const descriptionNode =
            detailsCardRef.current.querySelector<HTMLElement>('.line-clamp-2');

          if (
            prevState === 'TRUNCATE' &&
            (!descriptionNode || descriptionNode.offsetHeight <= descriptionNode.scrollHeight)
          ) {
            return 'NONE';
          } else if (
            prevState === 'FULL' &&
            DETAILS_CARD_DEFAULT_HEIGHT >= detailsCardRef.current.getBoundingClientRect().height
          ) {
            return 'NONE';
          } else if (
            prevState !== 'TRUNCATE' &&
            DETAILS_CARD_DEFAULT_HEIGHT < detailsCardRef.current.getBoundingClientRect().height
          ) {
            return 'TRUNCATE';
          }
          return prevState;
        });
      },
      1000,
      false,
    ),
    [setTruncateState, detailsCardRef],
  );

  useIsomorphicLayoutEffect(() => {
    global.addEventListener('resize', handleResize);
    return () => {
      global.removeEventListener('resize', handleResize);
    };
  }, []);

  return { truncateState, toggleTruncateState, detailsCardRef };
};
