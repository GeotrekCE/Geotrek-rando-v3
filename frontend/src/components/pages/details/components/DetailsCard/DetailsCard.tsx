import { useContext } from 'react';
import { textEllipsisAfterNLines } from 'services/cssHelpers';
import parse from 'html-react-parser';
import { HtmlText } from 'components/pages/details/utils';
import styled from 'styled-components';
import { FormattedMessage } from 'react-intl';
import { Link } from 'components/Link';
import { ListAndMapContext } from 'modules/map/ListAndMapContext';
import { CardIcon } from 'components/CardIcon';
import { getSpacing, MAX_WIDTH_MOBILE } from 'stylesheet';
import { useDetailsCard } from './useDetailsCard';
import { DetailsCardCarousel } from '../DetailsCardCarousel';
export interface DetailsCardProps {
  id: string;
  name: string;
  place?: string;
  description?: string;
  thumbnailUris: string[];
  iconUri?: string;
  logoUri?: string;
  className?: string;
  redirectionUrl?: string;
}

export const DetailsCard: React.FC<DetailsCardProps> = ({
  id,
  name,
  description,
  thumbnailUris,
  iconUri,
  place,
  logoUri,
  className,
  redirectionUrl,
}) => {
  const { truncateState, toggleTruncateState, heightState, detailsCardRef } = useDetailsCard();
  const descriptionStyled =
    truncateState === 'TRUNCATE' ? (
      <TruncatedHtmlText className="text-greyDarkColored">
        {parse(description ?? '')}
      </TruncatedHtmlText>
    ) : (
      <HtmlText className="text-greyDarkColored">{parse(description ?? '')}</HtmlText>
    );

  const { setHoveredCardId } = useContext(ListAndMapContext);

  return (
    <DetailsCardContainer
      height={heightState}
      className={`border border-solid border-greySoft rounded-card
      flex-none overflow-hidden relative
      flex flex-col h-auto desktop:flex-row desktop:w-auto mx-1
      cursor-pointer hover:border-blackSemiTransparent transition-all duration-500
      ${className ?? ''}`}
      onMouseEnter={() => {
        setHoveredCardId(id);
      }}
      onMouseLeave={() => {
        setHoveredCardId(null);
      }}
    >
      {logoUri && (
        <img
          className="hidden desktop:absolute h-12 object-cover object-center right-6 top-6"
          src={logoUri}
        />
      )}
      <div className="flex-none desktop:w-2/5">
        {thumbnailUris.length > 1 ? (
          <DetailsCardCarousel thumbnailUris={thumbnailUris} height={heightState} />
        ) : (
          <CardSingleImage src={thumbnailUris[0]} height={heightState} />
        )}
        {iconUri && <CardIcon iconUri={iconUri} />}
      </div>
      <div
        ref={detailsCardRef}
        className={`flex flex-col relative
        p-2 desktop:p-6 desktop:my-auto`}
      >
        {place && (
          <OptionalLink redirectionUrl={redirectionUrl}>
            <p className="text-greyDarkColored">{place}</p>
          </OptionalLink>
        )}
        <OptionalLink redirectionUrl={redirectionUrl}>
          <p className="text-Mobile-C1 desktop:text-H4 text-primary1 font-bold">{name}</p>
        </OptionalLink>
        {description && (
          <div
            className="mt-1 desktop:mt-4
            flex flex-col desktop:flex-row desktop:items-end
            text-Mobile-C2 desktop:text-P1 text-greyDarkColored"
          >
            <OptionalLink redirectionUrl={redirectionUrl}>{descriptionStyled}</OptionalLink>
            <span
              className="text-primary1 underline cursor-pointer flex-shrink-0 desktop:ml-1"
              onClick={toggleTruncateState}
            >
              <FormattedMessage
                id={truncateState === 'TRUNCATE' ? 'details.readMore' : 'details.readLess'}
              />
            </span>
          </div>
        )}
      </div>
    </DetailsCardContainer>
  );
};

interface OptionalLinkProps {
  redirectionUrl?: string;
  children: React.ReactNode;
}

const OptionalLink: React.FC<OptionalLinkProps> = ({ redirectionUrl, children }) => {
  return redirectionUrl !== undefined ? (
    <Link href={redirectionUrl}>{children}</Link>
  ) : (
    <>{children}</>
  );
};

const DetailsCardContainer = styled.div<{ height: number }>`
  height: fit-content;
  flex: none;
  overflow: hidden;
  display: flex;
  // Fix for border radius + overflow hidden in Safari, see https://gist.github.com/ayamflow/b602ab436ac9f05660d9c15190f4fd7b
  -webkit-mask-image: -webkit-radial-gradient(white, black);
  mask-image: radial-gradient(white, black);
  position: relative;
  flex-direction: column;
  @media (min-width: ${MAX_WIDTH_MOBILE}px) {
    height: ${props => props.height}px;
    width: auto;
    flex-direction: row;
    margin-bottom: ${getSpacing(6)};
  }
`;

export const CardSingleImage = styled.img<{ height: number }>`
  height: ${getSpacing(50)};
  width: ${getSpacing(60)};
  object-fit: cover;
  object-position: top;
  @media (min-width: ${MAX_WIDTH_MOBILE}px) {
    height: ${props => props.height}px;
    width: 100%;
  }
`;

const TruncatedHtmlText = styled(HtmlText)`
  ${textEllipsisAfterNLines(2)}
`;
