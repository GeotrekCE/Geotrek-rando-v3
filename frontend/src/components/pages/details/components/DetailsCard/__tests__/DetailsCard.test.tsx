import { render } from 'services/testing/reactTestingLibraryWrapper';
import { DetailsCard } from '../DetailsCard';
describe('DetailsCard', () => {
  it('should display a well parsed description Element', () => {
    const propsCard = {
      name: 'Église St Louis',
      id: '2',
      attachments: [
        {
          url: 'https://cdn.pixabay.com/photo/2017/06/29/18/40/background-2455710_1280.jpg',
          author: 'Lorem ipsum',
          legend: 'Lorem ipsum',
        },
      ],
      thumbnailUris: ['https://cdn.pixabay.com/photo/2017/06/29/18/40/background-2455710_1280.jpg'],
      iconUri: 'https://geotrekdemo.ecrins-parcnational.fr/media/upload/practice-foot_GpBv9u1.svg',
      iconName: 'Randonnée pédestre',
      description:
        "<span>Pour esp&eacute;rer apercevoir cet oiseau, partir la nuit au printemps, parcourir un grand d&eacute;nivel&eacute; afin d'arriver sur son terrain de pr&eacute;dilection &agrave; plus de 2000 m voire 3000 m d'altitude avant le lever du jour et l&agrave;, entendre le chant guttural&nbsp;caract&eacute;ristique qui trahit sa pr&eacute;sence. Mais pour le voir, il faudra bien ouvrir les yeux ou se munir d'une paire de jumelles. Et alors l&agrave;, quel bonheur&nbsp;! Le lagop&egrave;de alpin est l'esp&egrave;ce arctique par excellence, menac&eacute;e entre autre par le r&eacute;chauffement climatique. Il fait partie des esp&egrave;ces &agrave; prot&eacute;ger dans le c&oelig;ur du Parc national des Ecrins.</span>",
    };
    const detailsDescription = render(<DetailsCard {...propsCard} />);
    const detailsDescriptionWithCarousel = render(
      <DetailsCard
        {...propsCard}
        thumbnailUris={[
          'https://cdn.pixabay.com/photo/2017/06/29/18/40/background-2455710_1280.jpg',
          'https://cdn.pixabay.com/photo/2017/06/29/18/40/background-2455710_1280.jpg',
        ]}
      />,
    );

    expect(detailsDescription).toMatchSnapshot();
    expect(detailsDescriptionWithCarousel).toMatchSnapshot();
  });
});
