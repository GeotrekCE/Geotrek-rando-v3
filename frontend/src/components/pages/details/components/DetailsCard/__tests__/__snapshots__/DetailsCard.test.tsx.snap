// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailsCard should display a well parsed description Element 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="DetailsCard__DetailsCardContainer-sc-1paweiv-0 jZvBUu border border-solid border-greySoft rounded-card
      flex-none overflow-hidden relative
      flex flex-col h-auto desktop:flex-row desktop:w-auto mx-1
      cursor-pointer hover:border-blackSemiTransparent transition-all duration-500
      "
        height="200"
      >
        <div
          class="flex-none desktop:w-2/5"
        >
          <div
            style="background-color: white; position: relative;"
          >
            <div
              style="display: flex; align-items: center; justify-content: center; width: 100%; height: 100%;"
            >
              <div
                style="width: 100%; height: 100%;"
              >
                <div
                  class="slick-slider Carousel__StyledSlider-sc-1tck7ng-0 iyRBSY slick-initialized"
                >
                  
                  <div
                    class="slick-list"
                  >
                    <div
                      class="slick-track"
                      style="opacity: 1; transform: translate3d(0px, 0px, 0px);"
                    >
                      <div
                        aria-hidden="false"
                        class="slick-slide slick-active slick-current"
                        data-index="0"
                        style="outline: none; width: 0px;"
                        tabindex="-1"
                      >
                        <div>
                          <img
                            class="DetailsCard__CardSingleImage-sc-1paweiv-1 fzeiag"
                            height="200"
                            src="https://cdn.pixabay.com/photo/2017/06/29/18/40/background-2455710_1280.jpg"
                            style="width: 100%; display: inline-block;"
                            tabindex="-1"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                  
                  
                </div>
              </div>
            </div>
          </div>
          <div
            class="CardIcon__Wrapper-sc-cv6h96-0 iuMpeF absolute top-4 left-4 h-8 flex items-center w-auto rounded-full shadow-sm text-white border-2 border-white border-solid"
            color="#AA397D"
          >
            <div
              class="CardIcon__Label-sc-cv6h96-1 ddMeYP"
            >
              <div />
            </div>
          </div>
        </div>
        <div
          class="flex flex-col relative
        p-2 desktop:p-6 desktop:my-auto"
        >
          <p
            class="text-Mobile-C1 desktop:text-H4 text-primary1 font-bold"
          >
            Église St Louis
          </p>
          <div
            class="mt-1 desktop:mt-4 flex flex-col desktop:flex-row desktop:items-end text-Mobile-C2 desktop:text-P1 text-greyDarkColored"
          >
            <div
              class="utils__HtmlText-sc-16l79i-0 DetailsCard__TruncatedHtmlText-sc-1paweiv-2 jgscmn ewDPit text-greyDarkColored"
            >
              <span>
                Pour espérer apercevoir cet oiseau, partir la nuit au printemps, parcourir un grand dénivelé afin d'arriver sur son terrain de prédilection à plus de 2000 m voire 3000 m d'altitude avant le lever du jour et là, entendre le chant guttural caractéristique qui trahit sa présence. Mais pour le voir, il faudra bien ouvrir les yeux ou se munir d'une paire de jumelles. Et alors là, quel bonheur ! Le lagopède alpin est l'espèce arctique par excellence, menacée entre autre par le réchauffement climatique. Il fait partie des espèces à protéger dans le cœur du Parc national des Ecrins.
              </span>
            </div>
            <span
              class="text-primary1 underline cursor-pointer flex-shrink-0 desktop:ml-1"
            >
              lire la suite
            </span>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="DetailsCard__DetailsCardContainer-sc-1paweiv-0 jZvBUu border border-solid border-greySoft rounded-card
      flex-none overflow-hidden relative
      flex flex-col h-auto desktop:flex-row desktop:w-auto mx-1
      cursor-pointer hover:border-blackSemiTransparent transition-all duration-500
      "
      height="200"
    >
      <div
        class="flex-none desktop:w-2/5"
      >
        <div
          style="background-color: white; position: relative;"
        >
          <div
            style="display: flex; align-items: center; justify-content: center; width: 100%; height: 100%;"
          >
            <div
              style="width: 100%; height: 100%;"
            >
              <div
                class="slick-slider Carousel__StyledSlider-sc-1tck7ng-0 iyRBSY slick-initialized"
              >
                
                <div
                  class="slick-list"
                >
                  <div
                    class="slick-track"
                    style="opacity: 1; transform: translate3d(0px, 0px, 0px);"
                  >
                    <div
                      aria-hidden="false"
                      class="slick-slide slick-active slick-current"
                      data-index="0"
                      style="outline: none; width: 0px;"
                      tabindex="-1"
                    >
                      <div>
                        <img
                          class="DetailsCard__CardSingleImage-sc-1paweiv-1 fzeiag"
                          height="200"
                          src="https://cdn.pixabay.com/photo/2017/06/29/18/40/background-2455710_1280.jpg"
                          style="width: 100%; display: inline-block;"
                          tabindex="-1"
                        />
                      </div>
                    </div>
                  </div>
                </div>
                
                
              </div>
            </div>
          </div>
        </div>
        <div
          class="CardIcon__Wrapper-sc-cv6h96-0 iuMpeF absolute top-4 left-4 h-8 flex items-center w-auto rounded-full shadow-sm text-white border-2 border-white border-solid"
          color="#AA397D"
        >
          <div
            class="CardIcon__Label-sc-cv6h96-1 ddMeYP"
          >
            <div />
          </div>
        </div>
      </div>
      <div
        class="flex flex-col relative
        p-2 desktop:p-6 desktop:my-auto"
      >
        <p
          class="text-Mobile-C1 desktop:text-H4 text-primary1 font-bold"
        >
          Église St Louis
        </p>
        <div
          class="mt-1 desktop:mt-4 flex flex-col desktop:flex-row desktop:items-end text-Mobile-C2 desktop:text-P1 text-greyDarkColored"
        >
          <div
            class="utils__HtmlText-sc-16l79i-0 DetailsCard__TruncatedHtmlText-sc-1paweiv-2 jgscmn ewDPit text-greyDarkColored"
          >
            <span>
              Pour espérer apercevoir cet oiseau, partir la nuit au printemps, parcourir un grand dénivelé afin d'arriver sur son terrain de prédilection à plus de 2000 m voire 3000 m d'altitude avant le lever du jour et là, entendre le chant guttural caractéristique qui trahit sa présence. Mais pour le voir, il faudra bien ouvrir les yeux ou se munir d'une paire de jumelles. Et alors là, quel bonheur ! Le lagopède alpin est l'espèce arctique par excellence, menacée entre autre par le réchauffement climatique. Il fait partie des espèces à protéger dans le cœur du Parc national des Ecrins.
            </span>
          </div>
          <span
            class="text-primary1 underline cursor-pointer flex-shrink-0 desktop:ml-1"
          >
            lire la suite
          </span>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
